sklearn/ensemble/_hist_gradient_boosting/loss.py

```diff
@@ -52,24 +71,32 @@ def init_gradients_and_hessians(self, n_samples, prediction_dim):
         """
         shape = (prediction_dim, n_samples)
         gradients = np.empty(shape=shape, dtype=G_H_DTYPE)
+        if sample_weight is not None:
+            self.hessians_are_constant = False
```

Comment:
The `hessians_are_constant` is a bit awkward now.

hessians_are_constant is a class attribute, but here you're setting an instance attribute. So with sample_weights, we would have `LeastSquares.hessians_are_constant` to True but ` least_squares_loss_instance.hessians_are_constant` to False.

I'm not sure how to best do that though. Maybe @ogrisel will have an idea?

Reply:
I'm still feel like we could do something better here. @ogrisel ?

Reply:
Can we have `hessians_are_constant` defined in an `__init__` something like this:

```py
class BaseLoss:
    def __init__(self, hessians_are_constant):
        self.hessians_are_constant = hessians_are_constant

class LeastSquares(BaseLoss):
    def __init__(self, sample_weights=None):
		hessians_are_constant = True
        if sample_weights is not None:
            hessians_are_constant = False
        super().__init__(self, 
						 hessians_are_constant=hessians_are_constant)
```

And then in `BaseHistGradientBoosting._get_loss`:

````
def _get_loss(self, sample_weights):
    return _LOSSES[self.loss](sample_weights=sample_weights) 
```

Reply:
I agree. Marking as resolved.